api/src/db/schemaUpgrade/v3.ts

```diff
@@ -0,0 +1,76 @@
+import { PostDto } from "../../dto/PostDto";
+import { DbService } from "../db.service";
+import { TagDto } from "../../dto/TagDto";
+import { ImageDto } from "../../dto/ImageDto";
+import { ImageUploadDto } from "../../dto/ImageUploadDto";
+import { S3Service } from "../../s3/s3.service";
+import { processChangeRequest } from "../../changeRequests/processChangeRequest";
+import { ChangeReqDto } from "../../dto/ChangeReqDto";
+import { GroupDto } from "src/dto/GroupDto";
+
+/**
+ * Upgrade the database schema from version 2 to 3
+ * Update image field to imageData field in PostDto and TagDto documents and upload images to S3
+ */
+export default async function (db: DbService, s3: S3Service) {
+    const schemaVersion = await db.getSchemaVersion();
+    if (schemaVersion == 2) {
+        console.info("Upgrading database schema from version 2 to 3");
+        const groupQuery = await db.getGroups();
+        const groupIds = groupQuery.docs.map((group: GroupDto) => group._id);
+
+        await db.processAllDocs(async (doc: any) => {
+            if (!doc) {
+                return;
+            }
+            if (doc.error) {
+                console.error(`Unable to parse document: ${doc.parseError}`);
+                return;
+            }
+            if (doc.type && (doc.type == "post" || doc.type == "tag")) {
+                const d: PostDto | TagDto = doc;
```

Comment:
I feel this variable should be more descriptive, or maybe a comment above it stating that it is used to make sure that Doc is PostDto | TagDto, or that it's just a placeholder to not directly edit the "doc" variable.

Reply:
I need to check if the document is of type "Post" or "Tag". The first `doc.type` in the `if` statement is probably there to keep TypeScript happy as it would have worked in vanilla JS.

Reply:
I am actually talking about the naming of the "d" variable. But maybe it being a single character already signifies that it is a copied version of the current "doc". The rest of the code is nice and descriptive.